gg.1
.\" Copyright (c) 2021-2024 Omar Polo <op@omarpolo.com>
.\"
.\" Permission to use, copy, modify, and distribute this software for any
.\" purpose with or without fee is hereby granted, provided that the above
.\" copyright notice and this permission notice appear in all copies.
.\"
.\" THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
.\" WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
.\" MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
.\" ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
.\" WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
.\" ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
.\" OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
.Dd $Mdocdate: May 29 2024$
.Dt GG 1
.Os
.Sh NAME
.Nm gg
.Nd gemini client
.Sh SYNOPSIS
.Nm
.Bk -words
.Op Fl 23Nnq
.Op Fl C Ar cert
.Op Fl d Ar mode
.Op Fl H Ar sni
.Op Fl K Ar key
.Op Fl P Ar host Ns Oo : Ns Ar port Oc
.Op Fl T Ar seconds
.Ar gemini://...
.Ek
.Sh DESCRIPTION
.Nm
.Pq gemini get
fetches the given gemini page and prints it to standard output.
.Pp
The options are as follows:
.Bl -tag -width Ds
.It Fl 2
Use TLSv1.2.
.It Fl 3
Use TLSv1.3.
.It Fl C Ar certificate
Use the given client
.Ar certificate .
.It Fl d Ar mode
Specify what
.Nm
should print.
.Ar mode
can be one of:
.Bl -tag -width header -compact
.It Ic none
print only the body of the reply, the default.
.It Ic code
print only the response code.
.It Ic header
print only the response header.
.It Ic meta
print only the response meta.
.It Ic all
print the whole response as-is.
.El
.It Fl H Ar sni
Use the given
.Ar sni
host name instead of the one deducted by the IRI or proxy.
.It Fl K Ar key
Specify the key for the certificate.
It's mandatory if
.Fl C
is used.
.It Fl N
Disables the server name verification.
.It Fl n
Check the given IRI for validity, but don't issue any requests.
.It Fl P Ar host Ns Oo : Ns Ar port Oc
Connect to the given
.Ar host
and
.Ar port
to do the request instead of the ones extracted by the IRI.
.Ar port
is by default 1965.
IPv6 addresses have to be wrapped in braces, for e.g. [::1].
.It Fl q
Don't print server error messages to standard error.
.It Fl T Ar seconds
Kill
.Nm
after
.Ar seconds .
.El
.Sh EXIT STATUS
The
.Nm
utility exits with zero if the response code was in the 2x range.
If a failure occurs, it exits with status code 1.
Otherwise, the error code reflects the Gemini response code.
.Sh ACKNOWLEDGEMENTS
.Nm
uses the
.Dq Flexible and Economical
UTF-8 decoder written by
.An Bjoern Hoehrmann .
.Sh SEE ALSO
.Xr ftp 1 ,
.Xr titan 1
.Sh AUTHORS
.An -nosplit
The
.Nm
utility was written by
.An Omar Polo Aq Mt op@omarpolo.com .
.Sh CAVEATS
.Nm
doesn't perform TOFU
.Pq Trust On First Use
or any X.509 certificate validation beyond the name verification.
.Pp
.Nm
doesn't follow redirects.


ChangeLog
2024-08-03  Omar Polo  <op@omarpolo.com>

	* configure (VERSION): release 2.1 "From Mars To Sirius"
	* add support for the proxy-protocol v1 in outgoing connections too;
	needs to be enabled with the new `proxy-v1' flag in a `proxy' block.
	* gg.c (parse_proxy): support IPv6 addresses in -P

2024-06-17  Christoph Liebender  <cristoph.liebender@posteo.de>

	* add support for the proxy-protocol v1 on incoming connection;
	needs to be enabled with the new `proxy-v1' flag for `listen'.

2024-06-17  Omar Polo  <op@omarpolo.com>

	* titan.c (open_input_file): improve error checking; errors like
	"no space left on device" would have been hidden otherwise.

2024-06-14  Omar Polo  <op@omarpolo.com>

	* gmid.c: correctly propagate the prefork value; was causing
	failures if the `prefork' setting was changed from the default.

2024-06-12  Omar Polo  <op@omarpolo.com>

	* gmid.c (main): fix typo in a bounds check resulting in
	"chroot path too long" on some systems.

2024-06-11  Omar Polo  <op@omarpolo.com>

	* configure (VERSION): release 2.0.5

2024-06-10  Omar Polo  <op@omarpolo.com>

	* don't error on a '..' component at the start of the path
	* reject NUL bytes embedded in the request

2024-06-09  Omar Polo  <op@omarpolo.com>

	* check for truncation various strlcpy calls.
	* clean up of a few unused prototypes and externs.

2024-06-08  Omar Polo  <op@omarpolo.com>

	* configure: change how strnvis(3) is handled: on systems
	with the broken interface gmid will just use its built-in
	version.

2024-06-06  Omar Polo  <op@omarpolo.com>

	* parse.y: allow again empty lines at the start of the config
	* configure (VERSION): release 2.0.4
	* portability fix for system with a wrong strnvis(3)

2024-06-05  Omar Polo  <op@omarpolo.com>

	* parse.y: add a nicer error message if the removed `cgi' option
	is still used.

2024-06-04  Omar Polo  <op@omarpolo.com>

	* configure (VERSION): release 2.0.3

2024-06-03  Omar Polo  <op@omarpolo.com>

	* server.c (vhost_fastcgi): fix `fastcgi off' handling.
	Reported by Alex // nytpu

2024-05-29  Omar Polo  <op@omarpolo.com>

	* server.c (handle_handshake): relax the SNI requirement.  There are
	legitimate use-cases where SNI can't be used, like connecting to a
	raw IPv6 address.
	* gg.c (main): add -q to avoid printing "Server Says:"
	* gg.c (main): unbreak -n
	* iri.c (parse_authority): add support for raw IPv6 addresses

2024-04-04  Omar Polo  <op@omarpolo.com>

	* configure (VERSION): release 2.0.2

2024-04-03  Omar Polo  <op@omarpolo.com>

	* configure: improve function checking in the configure
	* have/landlock.c: fix landlock test
	* gmid.c (main_print_conf): fix config dumping with -nn

2024-03-03  Omar Polo  <op@omarpolo.com>

	* gmid.c: fix `log access path' with a chroot

2024-01-30  Anna “CyberTailor”

	* contrib/vim/indent/gmid.vim: fix indent

2024-01-30  Omar Polo  <op@omarpolo.com>

	* parse.y: don't make log styles reserved keywords.  Unbreaks the
	example in the manpage with `common = ...'.

2024-01-26  Omar Polo  <op@omarpolo.com>

	* parse.y: rework grammar to allow the semicolon after
	variables/macros definition and top-level options

2024-01-24  Omar Polo  <op@omarpolo.com>

	* configure (VERSION): release 2.0.1

2024-01-21  Omar Polo  <op@omarpolo.com>

	* convert gmid to the new imsg API

2024-01-14  Anna “CyberTailor”

	* configure: fix --mandir handling

2024-01-11  Omar Polo  <op@omarpolo.com>

	* configure (VERSION): release 2.0

2024-01-09  Anna “CyberTailor”

	* contrib/vim/syntax/gmid.vim: update Vim syntax file
	* contrib/vim/ale_listers/gmid/gmid.vim: add ALE linter

2023-10-18  Omar Polo  <op@omarpolo.com>

	* ge.c (load_local_cert): generate EC certs by default, use -R to keep generating RSA ones.
	* utils.c (gencert): generate EC too
	* gg.c (get): print the response header for non-2x replies to standard error

2023-10-15  Omar Polo  <op@omarpolo.com>

	* gg.c (main): exit with the gemini response code (unless it's 2x)

2023-08-29  Omar Polo  <op@omarpolo.com>

	* ge.c (data_dir): use $XDG_DATA_HOME/gemexp instead of /gmid for the certificates.

2023-08-18  Omar Polo  <op@omarpolo.com>

	* fcgi.c (fcgi_req): sync the parameters with RFC3875 (CGI)
	(fcgi_req): send "GET" as REQUEST_METHOD

2023-08-08  Omar Polo  <op@omarpolo.com>

	* fcgi.c (fcgi_req): implement SCRIPT_NAME / PATH_INFO splitting for fastcgi
	(fcgi_req): add `fastcgi strip'

2023-08-07  Omar Polo  <op@omarpolo.com>

	* logger.c (logger_dispatch_server): allow to change the syslog(3) facility.

	* gmid.c (main): attempt to load TLS certificates and load mimes and virtual hosts root as part of configtest (-n) rather than verifying the syntax of the configuration only.

2023-07-25  Omar Polo  <op@omarpolo.com>

	* gmid.c (log_request): allow to change the logging style.

2023-07-24  Omar Polo  <op@omarpolo.com>

	* parse.y: add ability to log to files with `log access <path>'.

2023-07-23  Omar Polo  <op@omarpolo.com>

	* parse.y: revamp fastcgi configuration; make it per-location

2023-07-22  Omar Polo  <op@omarpolo.com>

	* titan.c (main): add titan(1), a simple titan client

	* gg.c (get): warn when the server doesn't use TLS' close notify

2023-07-01  Omar Polo  <op@omarpolo.com>

	* fcgi.c (fcgi_handle_stdout): parse and log the fastcgi reply

2023-06-24  Omar Polo  <op@omarpolo.com>

	* server.c (handle_handshake): correctly handle TLS handshake failures.

	* server.c (client_close_ev): plug memory leak

2023-06-23  Omar Polo  <op@omarpolo.com>

	* parse.y: implement `listen on'

2023-06-13  Omar Polo  <op@omarpolo.com>

	* regress/sha: remove regress/sha; sha256/sha256sum is no more required for the regress suite.

2023-06-12  Omar Polo  <op@omarpolo.com>

	* regress/lib.sh (run_test): use the default prefork in tests

2023-06-11  Omar Polo  <op@omarpolo.com>

	* crypto.c: add a privsep crypto engine (enabled only on OpenBSD)

	* configure: add -Wpointer-sign to the mix, adjust the code to cope.

2023-05-05  Omar Polo  <op@omarpolo.com>


gg.c
/*
 * Copyright (c) 2021-2023 Omar Polo <op@omarpolo.com>
 *
 * Permission to use, copy, modify, and distribute this software for any
 * purpose with or without fee is hereby granted, provided that the above
 * copyright notice and this permission notice appear in all copies.
 *
 * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
 * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
 * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
 * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
 * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
 * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
 * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 */

#include "gmid.h"

#include <sys/socket.h>

#include <assert.h>
#include <ctype.h>
#include <errno.h>
#include <locale.h>
#include <string.h>
#include <wchar.h>

enum debug {
	DEBUG_NONE,
	DEBUG_CODE,
	DEBUG_HEADER,
	DEBUG_META,
	DEBUG_ALL,
};

/* flags */
int		 debug;
int		 dont_verify_name;
int		 flag2;
int		 flag3;
int		 nop;
int		 redirects = 5;
int		 timer;
int		 quiet;
const char	*cert;
const char	*key;
const char	*proxy_host;
const char	*proxy_port;
const char	*sni;

/* state */
struct tls_config *tls_conf;

static void
timeout(int signo)
{
	dprintf(2, "%s: timer expired\n", getprogname());
	exit(1);
}

static void
load_tls_conf(void)
{
	if ((tls_conf = tls_config_new()) == NULL)
		err(1, "tls_config_new");

	tls_config_insecure_noverifycert(tls_conf);
	if (dont_verify_name)
		tls_config_insecure_noverifyname(tls_conf);

	if (flag2 &&
	    tls_config_set_protocols(tls_conf, TLS_PROTOCOL_TLSv1_2) == -1)
		errx(1, "can't set TLSv1.2");
	if (flag3 &&
	    tls_config_set_protocols(tls_conf, TLS_PROTOCOL_TLSv1_3) == -1)
		errx(1, "can't set TLSv1.3");

	if (cert != NULL &&
	    tls_config_set_keypair_file(tls_conf, cert, key) == -1)
		errx(1, "can't load client certificate %s", cert);
}

static void
connectto(struct tls *ctx, const char *host, const char *port)
{
	struct addrinfo hints, *res, *res0;
	int error;
	int saved_errno;
	int s;
	const char *cause = NULL;
	const char *sname;

	if (proxy_host != NULL) {
		host = proxy_host;
		port = proxy_port;
	}

	if ((sname = sni) == NULL)
		sname = host;

	memset(&hints, 0, sizeof(hints));
	hints.ai_family = AF_UNSPEC;
	hints.ai_socktype = SOCK_STREAM;
	error = getaddrinfo(host, port, &hints, &res0);
	if (error)
		errx(1, "%s", gai_strerror(error));

	s = -1;
	for (res = res0; res != NULL; res = res->ai_next) {
		s = socket(res->ai_family, res->ai_socktype,
		    res->ai_protocol);
		if (s == -1) {
			cause = "socket";
			continue;
		}

		if (connect(s, res->ai_addr, res->ai_addrlen) == -1) {
			cause = "connect";
			saved_errno = errno;
			close(s);
			errno = saved_errno;
			s = -1;
			continue;
		}

		break;
	}

	if (s == -1)
		err(1, "%s: can't connect to %s:%s", cause,
		    host, port);

	freeaddrinfo(res0);

	if (tls_connect_socket(ctx, s, sname) == -1)
		errx(1, "tls_connect_socket: %s", tls_error(ctx));
}

static void
doreq(struct tls *ctx, const char *buf)
{
	size_t	s;
	ssize_t	w;

	s = strlen(buf);
	while (s != 0) {
		switch (w = tls_write(ctx, buf, s)) {
		case 0:
		case -1:
			errx(1, "tls_write: %s", tls_error(ctx));
		case TLS_WANT_POLLIN:
		case TLS_WANT_POLLOUT:
			continue;
		}

		s -= w;
		buf += w;
	}
}

static size_t
dorep(struct tls *ctx, uint8_t *buf, size_t len)
{
	ssize_t	w;
	size_t	tot = 0;

	while (len != 0) {
		switch (w = tls_read(ctx, buf, len)) {
		case 0:
			return tot;
		case -1:
			errx(1, "tls_write: %s", tls_error(ctx));
		case TLS_WANT_POLLIN:
		case TLS_WANT_POLLOUT:
			continue;
		}

		len -= w;
		buf += w;
		tot += w;
	}

	return tot;
}

static void
safeprint(FILE *fp, const char *str)
{
	int		 len;
	wchar_t		 wc;

	for (; *str != '\0'; str += len) {
		if ((len = mbtowc(&wc, str, MB_CUR_MAX)) == -1) {
			mbtowc(NULL, NULL, MB_CUR_MAX);
			fputc('?', fp);
			len = 1;
		} else if (wcwidth(wc) == -1) {
			fputc('?', fp);
		} else if (wc != L'\n')
			putwc(wc, fp);
	}

	fputc('\n', fp);
}

static int
get(const char *r)
{
	struct tls	*ctx;
	struct iri	 iri;
	int		 foundhdr = 0, code = -1, od;
	char		 iribuf[GEMINI_URL_LEN];
	char		 req[GEMINI_URL_LEN];
	uint8_t		 buf[2048];
	const char	*parse_err, *host, *port;
	int		 ret;

	if (strlcpy(iribuf, r, sizeof(iribuf)) >= sizeof(iribuf))
		errx(1, "iri too long: %s", r);

	ret = snprintf(req, sizeof(req), "%s\r\n", r);
	if (ret < 0 || (size_t)ret >= sizeof(req))
		errx(1, "iri too long: %s", r);

	if (!parse_iri(iribuf, &iri, &parse_err))
		errx(1, "invalid IRI: %s", parse_err);

	if (nop)
		errx(0, "IRI OK");

	if ((ctx = tls_client()) == NULL)
		errx(1, "can't create tls context");

	if (tls_configure(ctx, tls_conf) == -1)
		errx(1, "tls_configure: %s", tls_error(ctx));

	host = iri.host;
	port = "1965";
	if (*iri.port != '\0')
		port = iri.port;

	connectto(ctx, host, port);

	od = 0;
	while (!od) {
		switch (tls_handshake(ctx)) {
		case 0:
			od = 1;
			break;
		case -1:
			errx(1, "handshake: %s", tls_error(ctx));
		}
	}

	doreq(ctx, req);

	for (;;) {
		uint8_t	*t;
		size_t	 len;

		len = dorep(ctx, buf, sizeof(buf));
		if (len == 0)
			break;

		if (foundhdr) {
			write(1, buf, len);
			continue;
		}
		foundhdr = 1;

		if (memmem(buf, len, "\r\n", 2) == NULL)
			errx(1, "invalid reply: no \\r\\n");
		if (!isdigit((unsigned char)buf[0]) ||
		    !isdigit((unsigned char)buf[1]) ||
		    buf[2] != ' ')
			errx(1, "invalid reply: invalid response format");

		code = (buf[0] - '0') * 10 + buf[1] - '0';

		if (debug == DEBUG_CODE) {
			printf("%d\n", code);
			break;
		}

		if (debug == DEBUG_HEADER) {
			t = memmem(buf, len, "\r\n", 2);
			assert(t != NULL);
			*t = '\0';
			printf("%s\n", buf);
			break;
		}

		if (debug == DEBUG_META) {
			t = memmem(buf, len, "\r\n", 2);
			assert(t != NULL);
			*t = '\0';
			printf("%s\n", buf+3);
			break;
		}

		if (debug == DEBUG_ALL) {
			write(1, buf, len);
			continue;
		}

		/* skip the header */
		t = memmem(buf, len, "\r\n", 2);
		assert(t != NULL);
		if (code < 20 || code >= 30) {
			*t = '\0';
			if (!quiet) {
				fprintf(stderr, "Server says: ");
				/* skip return code */
				safeprint(stderr, buf + 3);
			}
		}
		t += 2; /* skip \r\n */
		len -= t - buf;
		write(1, t, len);
	}

	for (;;) {
		switch (tls_close(ctx)) {
		case TLS_WANT_POLLIN:
		case TLS_WANT_POLLOUT:
			continue;
		case -1:
			warnx("tls_close: %s", tls_error(ctx));
			/* fallthrough */
		default:
			tls_free(ctx);
			return code;
		}
	}
}

static void __attribute__((noreturn))
usage(void)
{
	fprintf(stderr, "version: " GG_STRING "\n");
	fprintf(stderr, "usage: %s [-23Nnq] [-C cert] [-d mode] [-H sni] "
	    "[-K key] [-P host[:port]]\n",
	    getprogname());
	fprintf(stderr, "          [-T seconds] gemini://...\n");
	exit(1);
}

static int
parse_debug(const char *arg)
{
	if (!strcmp(arg, "none"))
		return DEBUG_NONE;
	if (!strcmp(arg, "code"))
		return DEBUG_CODE;
	if (!strcmp(arg, "header"))
		return DEBUG_HEADER;
	if (!strcmp(arg, "meta"))
		return DEBUG_META;
	if (!strcmp(arg, "all"))
		return DEBUG_ALL;
	usage();
}

static void
parse_proxy(char *arg)
{
	char *at;

	proxy_host = arg;
	proxy_port = "1965";

	if (*proxy_host == '[') {
		if ((at = strchr(proxy_host, ']')) == NULL)
			errx(1, "invalid host: %s", proxy_host);
		proxy_host++;
		*at++ = '\0';
		if (*at == '\0')
			return;
		if (*at != ':')
			errx(1, "invalid port specification: %s", at);
	} else if ((at = strchr(proxy_host, ':')) == NULL)
		return;

	*at++ = '\0';
	proxy_port = at;
}

int
main(int argc, char **argv)
{
	int		 ch, code;
	const char	*errstr;

	setlocale(LC_CTYPE, "");

	while ((ch = getopt(argc, argv, "23C:d:H:K:nNP:qT:")) != -1) {
		switch (ch) {
		case '2':
			flag2 = 1;
			break;
		case '3':
			flag3 = 1;
			break;
		case 'C':
			cert = optarg;
			break;
		case 'd':
			debug = parse_debug(optarg);
			break;
		case 'H':
			sni = optarg;
			break;
		case 'K':
			key = optarg;
			break;
		case 'N':
			dont_verify_name = 1;
			break;
		case 'n':
			nop = 1;
			break;
		case 'P':
			parse_proxy(optarg);
			dont_verify_name = 1;
			break;
		case 'q':
			quiet = 1;
			break;
		case 'T':
			timer = strtonum(optarg, 1, 1000, &errstr);
			if (errstr != NULL)
				errx(1, "timeout is %s: %s",
				    errstr, optarg);
			signal(SIGALRM, timeout);
			alarm(timer);
			break;
		default:
			usage();
		}
	}
	argc -= optind;
	argv += optind;

	if (flag2 + flag3 > 1) {
		warnx("only -2 or -3 can be specified at the same time");
		usage();
	}

	if ((cert != NULL && key == NULL) ||
	    (cert == NULL && key != NULL)) {
		warnx("cert or key is missing");
		usage();
	}

	if (argc != 1)
		usage();

	load_tls_conf();

	signal(SIGPIPE, SIG_IGN);

#ifdef __OpenBSD__
	if (pledge("stdio inet dns", NULL) == -1)
		err(1, "pledge");
#endif

	code = get(*argv);
	if (code >= 20 && code < 30)
		return 0;
	return code;
}
